### PR TITLE
[Optimize Instructions] Simplify sign extentions (Part 2)

### PR DESCRIPTION
**Pull request by MaxGraey**
Similar to #4004 but for 32-bit integers

```rust
i32(x) << 24 >> 24   ==>   i32.extend8_s(x)
i32(x) << 16 >> 16   ==>   i32.extend16_s(x)
```

### CODE REVIEW

**Comment by MaxGraey** [comment]
Fuzzed `ITERATION: 10114`

**Comment by MaxGraey** [comment]
@kripken @tlively Could you take a look at this PR please?

File: test/lit/passes/optimize-instructions-sign-ext.wast [hunk 1]
```diff
@@ -62,4 +62,34 @@
...
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (i32.const 8)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $i32-sign-extentions (param $x i32)
+    (drop (i32.shr_s (i32.shl (local.get $x) (i32.const 24)) (i32.const 24)))
+    (drop (i32.shr_s (i32.shl (local.get $x) (i32.const 16)) (i32.const 16)))
+    (drop (i32.shr_s (i32.shl (local.get $x) (i32.const 8)) (i32.const 8))) ;; skip
+  )
```
Comment (kripken) [inline test/lit/passes/optimize-instructions-sign-ext.wast:94]: Please add a test where the constants do not match. Say 24 and 16 (so each is valid by itself).

Comment (kripken) [reply test/lit/passes/optimize-instructions-sign-ext.wast:94]: Also a test where everything is good to optimize except for the outer shift being unsigned.

Comment (MaxGraey) [reply test/lit/passes/optimize-instructions-sign-ext.wast:135]: done!

**Review by kripken — APPROVED** [review]
Thanks!

**Comment by MaxGraey** [comment]
Thanks for review!

**Pull request merged**